src/varpubs/hgvs_extractor.py

```diff
@@ -0,0 +1,78 @@
+import re
+from cyvcf2 import VCF
+
+
+def get_annotation_field_index(vcf: VCF, field: str) -> int:
+    """
+    Retrieves the index of a specified field (e.g., 'HGVS.p', 'Gene_Name') from the ANN column in the VCF header.
+
+    Args:
+        vcf (VCF): A parsed VCF object from cyvcf2.
+        field (str): The desired annotation field to find in the ANN description.
+
+    Returns:
+        int: The index of the given annotation field.
+
+    Raises:
+        RuntimeError: If 'ANN' or the specified field is not found.
+    """
+    for rec in vcf.header_iter():
+        info = rec.info()
+        if info.get("ID") == "ANN":
+            desc = info.get("Description", "")
+            if field in desc:
+                fields_str = desc.split("': '")[-1].rstrip("'")
+                fields = [f.strip() for f in fields_str.split("|")]
+                try:
+                    return fields.index(field)
+                except ValueError as e:
+                    raise RuntimeError(f"{field} not found in ANN header") from e
+    raise RuntimeError("ANN field not found in VCF header")
+
+
+def extract_hgvsp_from_vcf(vcf_path: str) -> list[str]:
+    """
+    Extracts all unique HGVS.p terms from the given VCF file and converts them into two formats:
+    - Full form (e.g., 'KRAS p.Gly12Cys')
+    - Short form (e.g., 'KRAS G12')
+
+    Filters out synonymous (silent) mutations.
+
+    Args:
+        vcf_path (str): Path to the annotated VCF file.
+
+    Returns:
+        list[str]: A sorted list of unique variant terms to be searched in PubMed.
+    """
+    vcf = VCF(vcf_path)
+    hgvsp_index = get_annotation_field_index(vcf, "HGVS.p")
+    gene_index = get_annotation_field_index(vcf, "Gene_Name")
+    term_set = set()
+
+    for record in vcf:
+        ann = record.INFO.get("ANN")
+        if ann:
+            for ann_entry in ann.split(","):
+                fields = ann_entry.split("|")
+                if len(fields) > max(hgvsp_index, gene_index):
+                    hgvsp = fields[hgvsp_index]
+                    gene = fields[gene_index]
+
+                    # Skip entries not starting with 'p.' (protein-level annotation)
+                    if not hgvsp.startswith("p."):
+                        continue
+
+                    # Skip synonymous mutations (e.g., p.Tyr516Tyr)
+                    aa_change = re.findall(r"[A-Za-z]+", hgvsp)
+                    if len(aa_change) == 2 and aa_change[0] == aa_change[1]:
+                        continue
+
+                    # Add full term: e.g., 'KRAS p.Gly12Cys'
+                    term_set.add(f"{gene} {hgvsp}")
+
+                    # Add short form: e.g., 'KRAS G12'
+                    match = re.match(r"p\.\D+(\d+)", hgvsp)
+                    if match:
+                        term_set.add(f"{gene} G{match.group(1)}")
```

Comment:
This won't work in any case. There are multiple amino acids starting with G, for example. One has to look up the proper short codes in a table, which can be part of the codebase (e.g. as a dict in this module): https://www.cup.uni-muenchen.de/ch/compchem/tink/as.html

Reply:
Also, I would search for G12 here, but also for the more specific G12C.